src/domain/mod.rs

```diff
@@ -0,0 +1,83 @@
+use unicode_segmentation::UnicodeSegmentation;
+use validator::validate_email;
+
+pub trait Parse<T>: Sized {
+    fn parse(self) -> Result<T, ParseError>;
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Failed to validate: {0}")]
+pub struct ParseError(String);
+
+#[derive(Debug)]
+pub struct Username(String);
+
+impl Parse<Username> for String {
+    fn parse(self) -> Result<Username, ParseError> {
+        if valid_length(&self, 1, 255) {
+            Ok(Username(self))
+        } else {
+            Err(ParseError(
+                "Username must be between 1 and 255 characters long.".into(),
+            ))
+        }
+    }
```

Comment:
```suggestion
        if !valid_length(&self, 1, 255) {
            Err(ParseError("Username must be between 1 and 255 characters long.".into()))   
        }
        if !valid_xxx {
            Err(xxx)   
        }
        if !valid_yyy {
            Err(yyy)   
        }
        Ok(Username(self))
    }
```
[IMO]
`if else`の両方で`return`する場合は`else`節を省くことが多いかも。
`else`節が複数分岐するときはsuggestionのように早期returnの形式を取ることが多いです。

可読性の問題なので、最終的には自分の好きな風で大丈夫です〜！
（他の言語だとこの辺りもlinter指定できたりするけど、Rustもできるのかな？）

Reply:
validator クレートの実装覗いてみたらおっしゃる通りの書き方になってました。
早期リターンの形式に変更してみます。

Reply:
> （他の言語だとこの辺りもlinter指定できたりするけど、Rustもできるのかな？）

できそう(?)
(https://rust-lang.github.io/rust-clippy/rust-1.56.0/index.html)

Reply:
@ryuma017 @ggg-mzkr 関数型信者としてここは反対しておきたい

Reply:
```
pub fn ensure_validated<V, F>(value: V, validate: F, message: &str) -> Result<(), ValidationError>
where
    F: FnOnce(V) -> bool,
```

みたいなの書いて

```
ensure_validated(username)?;
```
とかで解決しておきたい

Reply:
というかシンプルに match で良い気がする

```rust
match valid_length(&self, 1, 255) {
    true => Ok(Username(self))
    _ => Err(...),
}
```

Reply:
@ggg-mzkr 
こちらでついでに書き換えました🫡 [c9eddd7](https://github.com/ryuma017/twimi/pull/23/commits/c9eddd78b1b56a57ea28372924cd9f345eea3247)

Reply:
気づくの遅れました。

Reply:
Rust は式指向なので if-else でも else を外して早期リターンにはあまりしない

Reply:
僕的には今回は validate するのが length だけなので match や if else でも足りるけど、条件が加わったときに新たに追記しやすいのは早期リターンの形式なのかなと深読みしてました( そこまで考える必要ないのか(?) )

Reply:
> 関数型信者としてここは反対しておきたい

なるほど〜。ifの前に見えないreturnがあるのか。